Jenkins Server

Maanasa-Kotha opened the pull request:
Started Jenkins server recipe but doesn't work yet.

Commit:
Created a jenkins agent recipe

This is a rough draft and has not been tested.

Commit:
Added a recipe for the jenkins agent and made a branch for it.

Not tested, mostly outlined.

Commit:
Added jenkins server recipe and modified things to work with it.

The Berskfile was added for the supermarket jenkins cookbook dependency. The metadata file was changed for the same reason. Attributes were added to fulfill some that were referenced in the jenkins recipe. Kitchen.yml was modified to add the recipe to the run list.

nuclearsandwich commented on kitchen.yml:
```diff
@@ -16,6 +16,7 @@ suites:
     data_bags_path: "test/integration/data_bags"
     run_list:
       - recipe[ros_buildfarm::agent]
+      - recipe[ros_buildfarm::jenkins]
```
Instead of running the jenkins and agent recpies together in the agent tests we should create an additional test suite specifically for the jenkins recipe. You can take a peek at #7 which does this for the repo recipe.

nuclearsandwich commented on metadata.rb:
```diff
@@ -9,3 +9,5 @@
 issues_url 'https://github.com/nuclearsandwich/cookbook-ros-buildfarm/issues'
 source_url 'https://github.com/nuclearsandwich/cookbook-ros-buildfarm/'
+
+depends 'jenkins'
```
You should add a version constraint here. Probably best to stick with the current major version (so use a constraint that limits us to any 8.*) https://docs.chef.io/config_rb_metadata/#version-constraints

Commit:
Changed the appropriate attributes in the imported agent recipe.

Commit:
Fixed the home directory of the templates file destination.

Commit:
Changed the docker package to docker.io and properly coded the data bags for keys and passwords.

Commit:
Fixing the run list to create a new suite for jenkins.

Commit:
Added version number for jenkins dependency.

Commit:
Added the test suite for jenkins.

Commit:
Added the correct timezone as an attribute to allow kitchen test to run successfully.

Commit:
Added the directory for the jenkins configuration as code file because the template resource was throwing an error that it didn't exist.

Commit:
Adding java to the jenkins recipe since it would throw an error.

The jdk is installed before including the jenkins recipe because it needs java and it was throwing errors during kitchen test.

Commit:
Commented out nginx because it isn't done yet but the rest without it works.

nuclearsandwich commented on attributes/jenkins.rb:
```diff
@@ -0,0 +1,2 @@
+default['ros_buildfarm']['jenkins']['admin']['key'] = ''
```
I think this will likely end up being a data bag item rather than an attribute since it is sensitive data.

nuclearsandwich commented on recipes/jenkins.rb:
```diff
@@ -0,0 +1,62 @@
+
+directory '/var/lib/jenkins/casc_configs'
+template '/var/lib/jenkins/casc_configs/jenkins.yaml' do
+  group node['jenkins']['master']['group']
```
Is there a reason you didn't update the file owner as well?

nuclearsandwich committed:
Add authentication scripts.

The authentication setup is rather heavy at first glance. We want to
enable both quick and standard Jenkins users with usernames and
passwords as well as other authorization / authentication strategies
such as the GitHub-delegated authentication used on the ROS build farms
and the delegated authorization used on ci.ros2.org (admin access is
granted to members of a specific team).

The default strategy reads user information from a data bag
'ros_buildfarm_jenkins_users' and uses the Project-based Matrix
Authorization strategy to set up defined permissions for users and
enabling anonymous read access.

As it is we're using an attribute to determine whether this 'default'
mode should be used or a second strategy which expects to find a Groovy
script for configuring authentication/authorization in a data bag
'ros_buildfarm_jenkins_scripts'. This strategy is used when the
attribute is set to 'groovy'.

nuclearsandwich committed:
Use plugin versions from build.ros2.org.

The previous plugin list was copied from ci.ros2.org incorrectly.

I've commented out the custom builds made by cottsay, before production
day we'll need to host those builds somewhere accessible to the internet
so they can be pulled in by url.

nuclearsandwich committed:
Run an apt update to start the recipe.

The apt_update resource will run an apt update if it has been 24 hours
since the last one. I'm always a little rankled that this isn't embedded
into all package resources but here we are.

nuclearsandwich committed:
Configure agent at the end of the recipe.

This is just cosmetic, it doesn't / should't actually affect what gets
built but it keeps the Jenkins server configuration front and center.

nuclearsandwich committed:
Restart jenkins if configuration file changes.

If this config file changes, we should restart Jenkins directly as
those changes may affect connect-able clients or cli tools which are
needed for the rest of the chef run.

This commit also adjusts the configuration path because as it turns out
the default path does not use the casc_configs directory but is in the
root of JENKINS_HOME, which is /var/lib/jenkins for the Debian package.

nuclearsandwich committed:
Pin to chef 15.

The require_chef_omnibus option is deprecated but the new one doesn't
seem to work. So do both for now.

nuclearsandwich committed:
Move kitchen file to make room for EC2 option.

nuclearsandwich committed:
Add an ec2-based Kitchen config.

This has a few extra setup steps including local AWS credentials. The
kitchen-ec2 gem has documented options:
https://github.com/test-kitchen/kitchen-ec2

nuclearsandwich committed:
Set security realm before creating users and fix permissions.

Jenkins starts not only with no authorization strategy but also no
security realm (as opposed to defaulting to Jenkins own user database).
I'm not strictly certain that the security realm needs to be in place
before user accounts are created but it seemed easy enought configure
authentication/authorization by
1. Establishing the security realm
2. Populating it with users.
3. Establishing the authorization strategy and permissions for those
   users.

Permissions weren't being set because I had an if .nil? which inverted
the intended condition check.
Also permissions were being mishandled because I reversed the pair order
in my little snippet so I fixed that.

nuclearsandwich committed:
Use http for cli access.

The current documentation for the jenkins cookbook paints SSH keys as
the easy way to manage the local CLI but Jenkins's ssh access functions
differently than the http access in that it binds to an address and port
rather than a port only and due to the way Jenkins is configured this
seems like it's going to be the public interface rather than the
private one. They're one and the same on our instances so if the ssh
connection is opened on the local interface it will work but the
SSH port is randomized and advertized via the X-Jenkins-SSH header or
somesuch and the CLI will try to use the public IP or domain which
means it leaves the security group and tries to re-enter it but the port
is blocked.

Also http is just as easy, the bug we monkey patched in our ros2ci
cookbook is resolved in the upstream cookbook for 8.0.

nuclearsandwich committed:
Whitespace fixes.

Not much going on here. Cleaning trailing whitespace from one line and
cleaning up an extra newline in the body.

nuclearsandwich committed:
Treat anonymous as a funny sort of user.

Instead of assuming that anyone using these recipes wants to enable
anonymous read access let them put permissions on an anonymous user but
don't create an anonymous user since I'm scared of shadowing the builtin
anonymous role.

nuclearsandwich committed:
Remove credentials from c-as-c since they'll be in data bags.

nuclearsandwich committed:
Remove security realm and authorization strategy from c-as-c.

These are both managed by chef directly so having them here would
override that.

nuclearsandwich committed:
Use Safe HTML markup formatter.

The Safe HTML markup formatter is actually preferred but none of our
config management captured this preference previously so our reference
instance didn't have the proper value set. I set it by hand and then
copied the generated c-as-c segment which gave me this result.

nuclearsandwich committed:
Views are managed by the ros_buildfarm library.

We might later add back the "Manage" view so we can set it as the
default.

nuclearsandwich committed:
Comment out warning parser for now as it's causing errors.

nuclearsandwich committed:
Replace some secrets with TODOs.

These will need to be templated based on data bag. Add todos to come
back to them shortly.

nuclearsandwich committed:
Add data bags for integration tests.

nuclearsandwich committed:
Remove unused java cookbook dependency.

nuclearsandwich committed:
Use a custom fork location for jenkins cookbook.

An additional change is required to handle retrying after auth failures
which needs to be contributed upstream. In the meantime lift the version
pin in favor of targeting a specific fork branch in Berkshelf.

nuclearsandwich committed:
Fix error in security realm script.

An operator precedence error was preventing this script from
successfully establishing the security realm causing errors on repeated
converge operations.

nuclearsandwich committed:
Clean up heredoc syntax.

Apparently this fancy >>~ heredoc style was added in Ruby 2.3 and will
strip leading whitespace from the heredoc without resorting to gsub
tricks at eval.

Also, the log entry was never seen so I just gave up on it.

nuclearsandwich committed:
Install acme.sh.

nuclearsandwich committed:
Restart jenkins after changing security realm.

nuclearsandwich committed:
Add webproxy with optional https support.

With this reconfiguration the webproxy (using nginx) will work with
https support via acme.sh and letsencrypt or without it using nginx as a
standard reverse proxy on port 80.

nuclearsandwich committed:
Apply fixes from cookstyle.

Running cookstyle caught a bunch of style nitpicks.
Hopefully to reduce git-annotate churn we'll start running it on commit
after this change.

You can ignore this commit when running git annotate with --ignore-rev
or adding it to a file with a list of revs and using --ignore-rev-file

nuclearsandwich commented:
@Maanasa-Kotha can you please update the PR description so it more closely describes the current state of the pull request. It would be a good idea to include a summary of what the recipe configures.

nuclearsandwich committed:
Add publish over ssh config.

The publish-over-ssh plugin doesn't work with configuration as code.
The same missing API for c-as-c also makes it unlikely to work with a
Groovy script so we have to fall back to the the xml configuration.

This is largely borrowed from a live config and uses a data bag for the
host configurations.

nuclearsandwich committed:
Rework ssh publisher configuration.

As we work on the configuration I've been doing my best not to copy
verbatim patterns from the previous buildfarm_deployment but instead to
closely adapt those that make sense while losely adapting or entirely
revisiting things which either didn't translate well to chef or things
which have held back approachability or configurability in the past.

Let down by the publish-over-ssh plugin's lack of support for
configuration-as-code and groovy scripting I think I got a little too
abstract trying to clean up and deduplicate template logic.  The
"no-such-thing-as-two" pattern I follow suggested that the four
different targets we have for publish-over-ssh may someday be three or
five and therefore storing bulk host configurations as data bags made
sense in the moment. However when I looked at what that asked of our
community build farm operators I hesitated and ultimately made this
change.

While there are N similar configurations and N may grow or shrink in
time. Right now N=4 are required in order for the build farm to operate
as expected. One of the challenges with the puppet-based buildfarm
deployment is the lack of delineation between hiera config values which
must or may be changed by individual operators and which ones are
"static" to the build farm and are in hiera only because there did not
seem to be a convenient other mechanism to configure them.

I saw in this previous strategy that we would need to ask operators to
create data bags with static values in them and that's something I would
like to avoid if at all possible. Hardening on this point may require
review of existing work to check for violations but it's at least on my
mind going forward.

Thus in reconciling this most recent feature I have pushed much of the
configuration into attributes, using a data bag only for the secret
which should definitely be operator supplied. As a result I've committed
what I consider to be a somewhat grave act of metaprogramming for a chef
recipe in the interest of Don't-Repeat-Yourself which I'm not entirely
convinced is the overriding concern. I'm hoping that the current
implementation will serve until I find a better way to express the
intent while avoiding duplicate changes.

While making this change, I also realized that the method we're using to
override attributes related to the agent recipe may not work when we try
for the single-host approach which was too difficult to achieve with the
existing buildfarm_deployment but which is not impossible in theory
though it will require more configuration that we've done today. So I
noted some TODOs related to that aspect in this change.

nuclearsandwich committed:
Add very basic smoke tests.

If jenkins or nginx aren't running something is definitely busted but
even if they're both running there could still be configuration
problems.

nuclearsandwich committed:
Add default server name attribute.

nuclearsandwich committed:
Update nginx default server handling.

I accidentally put the default_server designation in the server name
where it becomes an extra server name instead of on the listen directive
where it works as intended. Since the default config on Debian installs
a separate default server that config file needs to be removed.

I also dropped the .conf extension from the rendered template as it's
not used for the Debian style nginx sites.

nuclearsandwich committed:
Update copypasted comment in nginx config file.

nuclearsandwich committed:
Update SSH publisher config based on diff from Jenkins.

These changes, including an XML document version, the addition of a
proxy port option tag, a whitespace correction, and fixing a typo in the
plugin version attribute were all found comparing the generated file
with the one saved by Jenkins after a restart.

nuclearsandwich committed:
Move common config out of repo block and jenkins.yml

I think it's worth exploring trying to get more of the ssh publisher
config into the jenkins.yaml for now let's keep it together in XML but
try using the new style commonConfig.
